tt_metal/impl/device/device.hpp

```diff
@@ -396,12 +341,65 @@ class Device {
     std::tuple<SubDeviceManagerId, SubDeviceId> create_sub_device_manager_with_fabric(tt::stl::Span<const SubDevice> sub_devices, DeviceAddr local_l1_size);
     std::optional<SubDeviceId> get_fabric_sub_device_id() const;
 
-   private:
+    program_cache::detail::ProgramCache& get_program_cache() { return program_cache_; }
+
+    std::set<CoreCoord> get_compute_cores() const { return compute_cores_; }
+
+    uint32_t get_completion_queue_reader_core() const { return completion_queue_reader_core_; }
+
+    bool is_worker_queue_empty() const { return work_executor_.worker_queue.empty(); }
+
+    uint32_t get_trace_buffers_size() const { return trace_buffers_size_; }
+    void set_trace_buffers_size(uint32_t size) { trace_buffers_size_ = size; }
+
+    bool using_fast_dispatch() const { return using_fast_dispatch_; }
+
+    std::vector<std::vector<chip_id_t>> get_tunnels_from_mmio() const { return tunnels_from_mmio_; }
+
+    static constexpr MemoryAllocator allocator_scheme_ = MemoryAllocator::L1_BANKING;
+
+private:
     void initialize_default_sub_device_state(size_t l1_small_size, size_t trace_region_size, tt::stl::Span<const std::uint32_t> l1_bank_remap);
     SubDeviceManagerId get_next_sub_device_manager_id();
     void reset_sub_devices_state(const std::unique_ptr<detail::SubDeviceManager>& sub_device_manager);
     void MarkAllocationsUnsafe();
     void MarkAllocationsSafe();
+
+    chip_id_t id_;
+    uint32_t build_key_ = 0;
+    std::map<uint32_t, std::map<chip_id_t, std::vector<std::vector<std::tuple<tt_cxy_pair, dispatch_worker_build_settings_t>>>>> tunnel_device_dispatch_workers_;
```

Comment:
Woah, probably needs a named type...

Reply:
MapOfMapOfVectorOfVectorOfTuple ooh I never saw so large std type.
It should a class or map of classes. Type naming doesn't help here.

Reply:
> Type naming doesn't help here. 

🤣

yes